### ISSUE TITLE
NodePool ASG spans public & private subnets

### ISSUE THREAD

**Issue by swestcott** [coreos/kube-aws#389]
kube-aws v0.9.5-rc.1 (master also)

I have pre-provisioned subnets, gateways and routing tables. Controllers, workers & etcd are deployed to private subnets, with ELBs going in the public subnets. Bastion access is via peered VPC.

Relevant snippets from `cluster.yaml` below,

```
controller:
  autoScalingGroup:
    minSize: 1
    maxSize: 3
    rollingUpdateMinInstancesInService: 2
  subnets:
    - name: PrivateSubnet0
    - name: PrivateSubnet1
    - name: PrivateSubnet2
  loadBalancer:
    private: false
    subnets:
      - name: PublicSubnet0
      - name: PublicSubnet1
      - name: PublicSubnet2


worker:
  nodePools:
      name: nodepool1
        - name: PrivateSubnet0
        - name: PrivateSubnet1
        - name: PrivateSubnet2
      autoScalingGroup:
        minSize: 1
        maxSize: 3
        rollingUpdateMinInstancesInService: 2


subnets:
    - name: PublicSubnet0
      availabilityZone: eu-west-1a
      id: subnet-181e097c
    - name: PublicSubnet1
      availabilityZone: eu-west-1b
      id: subnet-fa7f6b8c
    - name: PublicSubnet2
      availabilityZone: eu-west-1c
      id: subnet-9cdd8dc4
    - name: PrivateSubnet0
      availabilityZone: eu-west-1a
      id: subnet-27faee43
    - name: PrivateSubnet1
      availabilityZone: eu-west-1b
      id: subnet-d72f3aa1
    - name: PrivateSubnet2
      availabilityZone: eu-west-1c
      id: subnet-3420706c

```

The issue is the resulting ASG configuration for the worker contains all public and private subnets,

![screen shot 2017-03-07 at 08 38 35](https://cloud.githubusercontent.com/assets/58629/23653220/b4eb76e8-0323-11e7-8c4f-fd0f44f754c0.png)

With the above configuration I'd expect workers to only be created in the private subnets.

**Comment by mumoshu** [comment]
Hi @swestcott!

The part of your configuration:
```yaml
worker:
  nodePools:
      name: nodepool1
        - name: PrivateSubnet0
        - name: PrivateSubnet1
        - name: PrivateSubnet2
```

should be:

```yaml
worker:
  nodePools:
  - name: nodepool1
     subnets:
      - name: PrivateSubnet0
      - name: PrivateSubnet1
      - name: PrivateSubnet2
```

`nodePools` is a array rather than a hash. Also, you need `subnets` under the node pool.

**Comment by swestcott** [comment]
```
worker:
  nodePools:
      name: nodepool1
      subnet:
        - name: PrivateSubnet0
        - name: PrivateSubnet1
        - name: PrivateSubnet2
      autoScalingGroup:
        minSize: 1
        maxSize: 3
        rollingUpdateMinInstancesInService: 2

**Comment by swestcott** [comment]
Apologies, I made a copy/paste mistake. Here's the raw configuration,

```yaml
worker:
  nodePools:
    - # Name of this node pool. Must be unique among all the node pools in this cluster
      name: nodepool1
#      # Number of worker nodes to create for an autoscaling group based pool
#      count: 3
#      # If omitted, public subnets are created by kube-aws and used for worker nodes
      subnet:
#      - # References subnets defined under the top-level `subnets` key by their names
        - name: PrivateSubnet0
        - name: PrivateSubnet1
        - name: PrivateSubnet2
#      # Auto Scaling Group definition for workers. If only `workerCount` is specified, min and max will be the set to that value and `rollingUpdateMinInstancesInService` will be one less.
      autoScalingGroup:
        minSize: 1
        maxSize: 3
        rollingUpdateMinInstancesInService: 2
```

**Comment by mumoshu** [comment]
@swestcott It should be `subnets` rather than `subnet`

**Comment by redbaron** [comment]
>  It should be subnets rather than subnet

looks like cluster.yaml grows complex enough so that failing validation on unknown fields can be justified

**Comment by swestcott** [comment]
@mumoshu Thanks, that fixed it. It looks like the typo in the template `cluster.yml` file has already been patched, fbc5b20b6e36f685017a3b49020bc055884a4552 👍

**Comment by c-knowles** [comment]
@redbaron know of anything decent for extra validation? I was looking around for the same thing but didn't come up with much other than some github issues in the golang project about JSON/XML parsing.

**Comment by mumoshu** [comment]
@c-knowles fyi, the extra validation is implemented in #425

**Comment by c-knowles** [comment]
Great, it's nice and I see it also pointed out some redundant and wrong config inside the tests!
